<a href="https://colab.research.google.com/github/bijeshsingha/NLP-Take-home-assignment/blob/master/NLP_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objective
The primary objective of this project is to develop and compare two robust machine learning models for sentiment analysis on a provided textual dataset. This comparative analysis will assess the impact of different text vectorization techniques: Bag-of-Words (BoW) and Term Frequency-Inverse Document Frequency (TF-IDF), on classification performance.

The execution will follow a detailed, multi-stage methodology:

* Data Preparation and NLP Cleansing: Load the dataset and apply a comprehensive suite of Natural Language Processing (NLP) text cleaning steps. This process includes standardizing text (lowercasing, removal of unwanted characters, URLs, and HTML tags), refining the corpus through effective handling of stopwords, correcting spelling errors, and replacing common acronyms with their full forms to ensure data quality and model readiness.

* Feature Engineering: Generate numerical feature representations of the cleaned text using two independent methods: CountVectorizer (BoW) and TfidfVectorizer (TF-IDF).

* Model Construction and Evaluation: Construct a classification model using the RandomForest algorithm over the feature sets generated by both the BoW and TF-IDF vectorizers.

* Comparative Analysis: Evaluate the performance of the two resulting classification models based on standard classification metrics (e.g., accuracy, precision, recall, F1-score). A critical analysis will be conducted to explain any observed differences in performance between the BoW-based and TF-IDF-based models, focusing on the theoretical advantages of each vectorization technique.


In [12]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("ankurzing/sentiment-analysis-for-financial-news")

print("Path to dataset files:", path)

Path to dataset files: /root/.cache/kagglehub/datasets/ankurzing/sentiment-analysis-for-financial-news/versions/5


In [3]:
!pip install pyspellchecker

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 39.0 MB/s eta 0:00:00


In [15]:
import pandas as pd
import sys

# Dictionary to store library versions
library_versions = {
    'pandas': 'N/A',
    'numpy': 'N/A',
    're': 'N/A', # re is a built-in module, usually doesn't have __version__
    'nltk': 'N/A',
    'pyspellchecker': 'N/A',
    'sklearn': 'N/A'
}

# Get versions of installed libraries
try:
    import pandas
    library_versions['pandas'] = pandas.__version__
except ImportError:
    pass

try:
    import numpy
    library_versions['numpy'] = numpy.__version__
except ImportError:
    pass

try:
    import nltk
    library_versions['nltk'] = nltk.__version__
except ImportError:
    pass

try:
    import spellchecker
    library_versions['pyspellchecker'] = spellchecker.__version__
except ImportError:
    pass

try:
    import sklearn
    library_versions['sklearn'] = sklearn.__version__
except ImportError:
    pass

# Create a DataFrame from the dictionary
versions_df = pd.DataFrame(list(library_versions.items()), columns=['Library', 'Version'])

# Display the DataFrame
display(versions_df)

Library Version
0          pandas   2.2.2
1           numpy   2.0.2
2              re     N/A
3            nltk   3.9.1
4  pyspellchecker   0.8.3
5         sklearn   1.6.1

In [13]:
import os
import pandas as pd

# List contents of the downloaded directory
file_path = '/root/.cache/kagglehub/datasets/ankurzing/sentiment-analysis-for-financial-news/versions/5'
print(os.listdir(file_path))

['FinancialPhraseBank', 'all-data.csv']


In [14]:
import pandas as pd

# Construct the full path to the CSV file
csv_file_path = os.path.join(file_path, 'all-data.csv')

# Load the CSV into a pandas DataFrame
df = pd.read_csv(csv_file_path, encoding='latin-1', header=None)

# Display the first 5 rows of the DataFrame
display(df.head())

0                                                  1
0   neutral  According to Gran , the company has no plans t...
1   neutral  Technopolis plans to develop in stages an area...
2  negative  The international electronic industry company ...
3  positive  With the new production plant the company woul...
4  positive  According to the company 's updated strategy f...

# Data Source Acknowledgment
The data utilized for this sentiment analysis project is the Financial News Sentiment Dataset, sourced from Kaggle.

Dataset Name: Sentiment Analysis for Financial News

Original Author/Contributor: Ankur Singh

Source URL: https://www.kaggle.com/datasets/ankurzing/sentiment-analysis-for-financial-news

The dataset consists of various news headlines and their corresponding sentiment labels (e.g., positive, negative, or neutral), making it highly suitable for the comparative analysis of NLP vectorization techniques and machine learning classification implemented in this project.


## Rename DataFrame Columns

### Subtask:
Rename the columns of the DataFrame 'df' to meaningful names like 'Sentiment' and 'Text' for better readability and usability.


**Reasoning**:
To improve readability and usability, I will rename the columns of the DataFrame `df` from `0` to 'Sentiment' and `1` to 'Text'.



In [16]:
df.rename(columns={0: 'Sentiment', 1: 'Text'}, inplace=True)
display(df.head())

Sentiment                                               Text
0   neutral  According to Gran , the company has no plans t...
1   neutral  Technopolis plans to develop in stages an area...
2  negative  The international electronic industry company ...
3  positive  With the new production plant the company woul...
4  positive  According to the company 's updated strategy f...

In [17]:
X = df['Text']
y = df['Sentiment']

print("X head:")
display(X.head())

print("\ny head:")
display(y.head())

X head:


0    According to Gran , the company has no plans t...
1    Technopolis plans to develop in stages an area...
2    The international electronic industry company ...
3    With the new production plant the company woul...
4    According to the company 's updated strategy f...
Name: Text, dtype: object


y head:


0     neutral
1     neutral
2    negative
3    positive
4    positive
Name: Sentiment, dtype: object

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (3876,)
Shape of X_test: (970,)
Shape of y_train: (3876,)
Shape of y_test: (970,)


In [19]:
print("type of X_train: {}".format( type(X_train)))

print(" length of text_train: {}".format(len( X_train)))

print("\n\n text_train[5]:\n{}".format( X_train[0:6]))

type of X_train: <class 'pandas.core.series.Series'>
 length of text_train: 3876


 text_train[5]:
432     In Russia , Raisio 's Food Division 's home ma...
3071    Operators only need to learn how to use one de...
289     The company expects its net sales in the first...
2588    The bridge will be 1.2 km long and is located ...
4121    Nokia and Capcom announced that Resident Evil ...
4489    At the same time I am delighted by the fact th...
Name: Text, dtype: object


In [20]:
print("type of X_test: {}".format( type(X_test)))

print(" length of X_test: {}".format(len(X_test)))

print("\n\n X_test[6]:\n{}".format( X_test.iloc[6]))

print("\n\n y_test[6]:\n{}".format (y_test.iloc[6]))

type of X_test: <class 'pandas.core.series.Series'>
 length of X_test: 970


 X_test[6]:
Finnish-owned contract manufacturer of electronics Elcoteq Hungary Kft has announced plans to recruit more than 650 new staffers to fulfill new orders in P+_cs , where the company has two plants .


 y_test[6]:
positive


#

# Text Cleaning

# Task
Define a Python function to clean text data by converting it to lowercase, removing URLs, stripping HTML tags, and removing all non-alphanumeric characters (except spaces) and extra spaces, then apply this function to the `X_train` and `X_test` Series.

## Define Text Cleaning Functions (URLs, HTML, Special Chars, Lowercase)

Create a function that takes a text string and performs several cleaning operations: converting to lowercase, removing URLs, stripping HTML tags, and removing all non-alphanumeric characters (except spaces) and extra spaces. This function will consolidate multiple text cleansing steps into one reusable block.


**Reasoning**:
I need to define a `clean_text` function that performs various text cleaning operations as specified in the instructions, including converting to lowercase, removing URLs, stripping HTML tags, removing non-alphanumeric characters, and handling spaces. I will use the `re` module for this.



In [21]:
import re

def clean_text(text):
    """Performs several text cleaning operations on the input string."""
    # 1. Convert to lowercase
    text = text.lower()

    # 2. Remove URLs
    text = re.sub(r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', '', text)

    # 3. Strip HTML tags
    text = re.sub(r'<.*?>', '', text)

    # 4. Remove all non-alphanumeric characters (except spaces)
    text = re.sub(r'[^a-zA-Z0-9 ]', '', text)

    # 5. Replace multiple spaces with a single space and strip leading/trailing spaces
    text = re.sub(r'\s+', ' ', text).strip()

    return text

# Test the function with a sample text
sample_text = "This is a sample text with a URL: https://example.com/page?id=123, HTML <b>tags</b>, special characters like @#$%^&*, and multiple spaces. <br /><br /> Another line!"
cleaned_sample_text = clean_text(sample_text)

print("Original Text:", sample_text)
print("Cleaned Text:", cleaned_sample_text)


Original Text: This is a sample text with a URL: https://example.com/page?id=123, HTML <b>tags</b>, special characters like @#$%^&*, and multiple spaces. <br /><br /> Another line!
Cleaned Text: this is a sample text with a url html tags special characters like and multiple spaces another line


**Reasoning**:
Now that the `clean_text` function is defined and tested, I will apply this function to the 'Text' columns of `X_train` and `X_test` to clean the actual training and testing data.



In [22]:
X_train_cleaned = X_train.apply(clean_text)
X_test_cleaned = X_test.apply(clean_text)

print("Original X_train head (first 5 entries):")
display(X_train.head())

print("\nCleaned X_train head (first 5 entries):")
display(X_train_cleaned.head())

print("\nOriginal X_test head (first 5 entries):")
display(X_test.head())

print("\nCleaned X_test head (first 5 entries):")
display(X_test_cleaned.head())

Original X_train head (first 5 entries):


432     In Russia , Raisio 's Food Division 's home ma...
3071    Operators only need to learn how to use one de...
289     The company expects its net sales in the first...
2588    The bridge will be 1.2 km long and is located ...
4121    Nokia and Capcom announced that Resident Evil ...
Name: Text, dtype: object


Cleaned X_train head (first 5 entries):


432     in russia raisio s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 km long and is located b...
4121    nokia and capcom announced that resident evil ...
Name: Text, dtype: object


Original X_test head (first 5 entries):


3207    The company was supposed to deliver machinery ...
1684    UNC Charlotte would also deploy SSH Tectia Con...
1044    In 2009 , Lee & Man had a combined annual prod...
4145    `` That 's a very high figure on the European ...
1538    In Finland , the corresponding service is Alma...
Name: Text, dtype: object


Cleaned X_test head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    unc charlotte would also deploy ssh tectia con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object

# Task
Define a Python function `correct_spellings` that takes a text string as input, corrects any spelling errors in it using the `SpellChecker` from the `pyspellchecker` library, and returns the corrected text.

## Define a Python function `correct_spellings` that takes a text string as input, corrects any spelling errors in it using the `SpellChecker` from the `pyspellchecker` library, and returns the corrected text.

### Subtask:
Define a function `correct_spellings` that uses `SpellChecker` to correct spelling errors in an input text string.


**Reasoning**:
I need to import the `SpellChecker` class, instantiate it, define the `correct_spellings` function as per the instructions, and then test it with a sample sentence.



In [23]:
from spellchecker import SpellChecker

# Instantiate a SpellChecker object (language='en' by default)
spell = SpellChecker(language='en')

def correct_spellings(text):
    """Corrects spelling errors in the input text using SpellChecker."""
    # Split the text into words
    words = text.split()
    corrected_words = []
    for word in words:
        # Get the most probable correct spelling for each word
        corrected_word = spell.correction(word)
        if corrected_word is not None:
            corrected_words.append(corrected_word)
        else:
            # If no correction is found, keep the original word
            corrected_words.append(word)

    # Join the corrected words back into a single string
    return " ".join(corrected_words)

# Test the function with a sample sentence
sample_text_with_errors = "This is a samle txt with some speling erors."
corrected_sample_text = correct_spellings(sample_text_with_errors)

print("Original Text:", sample_text_with_errors)
print("Corrected Text:", corrected_sample_text)

Original Text: This is a samle txt with some speling erors.
Corrected Text: This is a same text with some spelling errors


**Reasoning**:
Now that the `correct_spellings` function is defined and tested, I will apply this function to the `X_train_cleaned` and `X_test_cleaned` Series to correct spelling errors in the training and testing data.



In [24]:

X_train_spell_corrected = X_train_cleaned.apply(correct_spellings)
X_test_spell_corrected = X_test_cleaned.apply(correct_spellings)

print("Original X_train_cleaned head (first 5 entries):")
display(X_train_cleaned.head())

print("\nSpell-corrected X_train head (first 5 entries):")
display(X_train_spell_corrected.head())

print("\nOriginal X_test_cleaned head (first 5 entries):")
display(X_test_cleaned.head())

print("\nSpell-corrected X_test head (first 5 entries):")
display(X_test_spell_corrected.head())

Original X_train_cleaned head (first 5 entries):


432     in russia raisio s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 km long and is located b...
4121    nokia and capcom announced that resident evil ...
Name: Text, dtype: object


Spell-corrected X_train head (first 5 entries):


432     in russia raisin s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 am long and is located b...
4121    nokia and captor announced that resident evil ...
Name: Text, dtype: object


Original X_test_cleaned head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    unc charlotte would also deploy ssh tectia con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object


Spell-corrected X_test head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    una charlotte would also deploy ssh tactic con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object

In [25]:
X_train_spell_corrected.to_excel("X_train_spell_corrected.xlsx")
X_test_spell_corrected.to_excel("X_test_spell_corrected.xlsx")

# Punctuation Remove

In [31]:
import string

PUNCT_TO_REMOVE = string.punctuation    #list of standard python punctuations


def remove_punctuation(text):
    """custom function to remove the punctuation"""
    return text.translate(str.maketrans('', '', PUNCT_TO_REMOVE))   # tranlate "" to "" and delete all punctuations


text = "hurray!"
remove_punctuation(text)


#PUNCT_TO_REMOVE = string.punctuation   # !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
#str.maketrans('', '', PUNCT_TO_REMOVE)

'hurray'

In [32]:
X_train_no_punct = X_train_spell_corrected.apply(remove_punctuation)
X_test_no_punct = X_test_spell_corrected.apply(remove_punctuation)

print("Original X_train_spell_corrected head (first 5 entries):")
display(X_train_spell_corrected.head())

print("\nProcessed X_train_no_punct head (first 5 entries):")
display(X_train_no_punct.head())

print("\nOriginal X_test_spell_corrected head (first 5 entries):")
display(X_test_spell_corrected.head())

print("\nProcessed X_test_no_punct head (first 5 entries):")
display(X_test_no_punct.head())

Original X_train_spell_corrected head (first 5 entries):


432     in russia raisin s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 am long and is located b...
4121    nokia and captor announced that resident evil ...
Name: Text, dtype: object


Processed X_train_no_punct head (first 5 entries):


432     in russia raisin s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 am long and is located b...
4121    nokia and captor announced that resident evil ...
Name: Text, dtype: object


Original X_test_spell_corrected head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    una charlotte would also deploy ssh tactic con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object


Processed X_test_no_punct head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    una charlotte would also deploy ssh tactic con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object

# Remove stopwords

In [33]:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

# Convert to set
stop_words = set(ENGLISH_STOP_WORDS)

In [34]:
def remove_stopwords(text):
    """custom function to remove the stopwords"""

    return " ".join([word for word in str(text).split() if word not in stop_words])

In [35]:
X_train_no_stopwords = X_train_no_punct.apply(remove_stopwords)
X_test_no_stopwords = X_test_no_punct.apply(remove_stopwords)

print("Original X_train_no_punct head (first 5 entries):")
display(X_train_no_punct.head())

print("\nProcessed X_train_no_stopwords head (first 5 entries):")
display(X_train_no_stopwords.head())

print("\nOriginal X_test_no_punct head (first 5 entries):")
display(X_test_no_punct.head())

print("\nProcessed X_test_no_stopwords head (first 5 entries):")
display(X_test_no_stopwords.head())

Original X_train_no_punct head (first 5 entries):


432     in russia raisin s food division s home market...
3071    operators only need to learn how to use one de...
289     the company expects its net sales in the first...
2588    the bridge will be 12 am long and is located b...
4121    nokia and captor announced that resident evil ...
Name: Text, dtype: object


Processed X_train_no_stopwords head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operators need learn use device multiple level...
289     company expects net sales half 2008 reach prev...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Original X_test_no_punct head (first 5 entries):


3207    the company was supposed to deliver machinery ...
1684    una charlotte would also deploy ssh tactic con...
1044    in 2009 lee man had a combined annual producti...
4145    that s a very high figure on the european scal...
1538    in finland the corresponding service is alma m...
Name: Text, dtype: object


Processed X_test_no_stopwords head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma media s etu...
Name: Text, dtype: object

In [36]:
from nltk.stem import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()

def lemmatize_words(text):
    return " ".join([lemmatizer.lemmatize(word) for word in text.split()])


In [38]:
import nltk
nltk.download('wordnet')

X_train_lemmatized = X_train_no_stopwords.apply(lemmatize_words)
X_test_lemmatized = X_test_no_stopwords.apply(lemmatize_words)

print("Original X_train_no_stopwords head (first 5 entries):")
display(X_train_no_stopwords.head())

print("\nProcessed X_train_lemmatized head (first 5 entries):")
display(X_train_lemmatized.head())

print("\nOriginal X_test_no_stopwords head (first 5 entries):")
display(X_test_no_stopwords.head())

print("\nProcessed X_test_lemmatized head (first 5 entries):")
display(X_test_lemmatized.head())

[nltk_data] Downloading package wordnet to /root/nltk_data...


Original X_train_no_stopwords head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operators need learn use device multiple level...
289     company expects net sales half 2008 reach prev...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Processed X_train_lemmatized head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operator need learn use device multiple level ...
289     company expects net sale half 2008 reach previ...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Original X_test_no_stopwords head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma media s etu...
Name: Text, dtype: object


Processed X_test_lemmatized head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma medium s et...
Name: Text, dtype: object

In [39]:
!pip install demoji

import demoji

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.9/42.9 kB 1.2 MB/s eta 0:00:00


## Replace Emojis with Words

To ensure all text data is in a consistent and analyzable format, emojis will be replaced with their corresponding English word descriptions. This step helps in standardizing the text and preventing emojis from being treated as irrelevant characters during feature extraction.

In [40]:
def replace_emojis_with_words(text):
    """Replaces emojis in a string with their corresponding English word descriptions."""
    return demoji.replace_with_desc(text, sep=" ")

# Test the function with a sample text
sample_text_with_emojis = "Hello world! 👋 This is great 👍😊"
cleaned_sample_text_emojis = replace_emojis_with_words(sample_text_with_emojis)

print("Original Text:", sample_text_with_emojis)
print("Cleaned Text:", cleaned_sample_text_emojis)

Original Text: Hello world! 👋 This is great 👍😊
Cleaned Text: Hello world!  waving hand  This is great  thumbs up  smiling face with smiling eyes 


In [41]:
X_train_no_emojis = X_train_lemmatized.apply(replace_emojis_with_words)
X_test_no_emojis = X_test_lemmatized.apply(replace_emojis_with_words)

print("Original X_train_lemmatized head (first 5 entries):")
display(X_train_lemmatized.head())

print("\nProcessed X_train_no_emojis head (first 5 entries):")
display(X_train_no_emojis.head())

print("\nOriginal X_test_lemmatized head (first 5 entries):")
display(X_test_lemmatized.head())

print("\nProcessed X_test_no_emojis head (first 5 entries):")
display(X_test_no_emojis.head())

Original X_train_lemmatized head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operator need learn use device multiple level ...
289     company expects net sale half 2008 reach previ...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Processed X_train_no_emojis head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operator need learn use device multiple level ...
289     company expects net sale half 2008 reach previ...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Original X_test_lemmatized head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma medium s et...
Name: Text, dtype: object


Processed X_test_no_emojis head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma medium s et...
Name: Text, dtype: object

## Replace Acronyms with Full Forms

To standardize the text and improve readability, common acronyms will be replaced with their full word forms. This helps in consistent interpretation and feature extraction.

In [42]:
# Create a Python dictionary for common acronyms and their full forms
acronyms = {"AFAIK" : "As Far As I Know",
"AFK" : "Away From Keyboard",
"ASAP" : "As Soon As Possible",
"ATK" : "At The Keyboard",
"ATM" : "At The Moment",
"A3" : "Anytime, Anywhere, Anyplace",
"BAK" : "Back At Keyboard",
"BBL" : "Be Back Later",
"BBS" : "Be Back Soon",
"BFN" : "Bye For Now",
"B4N" : "Bye For Now",
"BRB" : "Be Right Back",
"BRT" : "Be Right There",
"BTW" : "By The Way",
"B4" : "Before",
"B4N" : "Bye For Now",
"CU" : "See You",
"CUL8R" : "See You Later",
"CYA" : "See You",
"FAQ" : "Frequently Asked Questions",
"FC" : "Fingers Crossed",
"FWIW" : "For What It's Worth",
"FYI" : "For Your Information",
"GAL" : "Get A Life",
"GG" : "Good Game",
"GN" : "Good Night",
"GMTA" : "Great Minds Think Alike",
"GR8" : "Great!",
"G9" : "Genius",
"IC" : "I See",
"ICQ" : "I Seek you",
"ILU" : "I Love You",
"IMHO" : "In My Honest/Humble Opinion",
"IMO" : "In My Opinion",
"IOW" : "In Other Words",
"IRL" : "In Real Life",
"KISS" : "Keep It Simple, Stupid",
"LDR" : "Long Distance Relationship",
"LMAO" : "Laugh My A.. Off",
"LOL" : "Laughing Out Loud",
"LTNS" : "Long Time No See",
"L8R" : "Later",
"MTE" : "My Thoughts Exactly",
"M8" : "Mate",
"NRN" : "No Reply Necessary",
"OIC" : "Oh I See",
"PITA" : "Pain In The A..",
"PRT" : "Party",
"PRW" : "Parents Are Watching",
"ROFL" : "Rolling On The Floor Laughing",
"ROFLOL" : "Rolling On The Floor Laughing Out Loud",
"ROTFLMAO" : "Rolling On The Floor Laughing My A.. Off",
"SK8" : "Skate",
"STATS" : "Your sex and age",
"ASL" : "Age, Sex, Location",
"THX" : "Thank You",
"TTFN" : "Ta-Ta For Now!",
"TTYL" : "Talk To You Later",
"U" : "You",
"U2" : "You Too",
"U4E" : "Yours For Ever",
"WB" : "Welcome Back",
"WTF" : "What The F...",
"WTG" : "Way To Go!",
"WUF" : "Where Are You From?",
"W8" : "Wait..."
}

def replace_acronyms(text):
    """Replaces common acronyms in a string with their full forms."""
    # Use a regular expression to find whole words that are acronyms
    # and replace them. re.IGNORECASE makes it case-insensitive.
    for acr, full_form in acronyms.items():
        text = re.sub(r'\b' + re.escape(acr) + r'\b', full_form, text, flags=re.IGNORECASE)
    return text

# Test the function with a sample text
sample_text_with_acronyms = "fyi, the meeting is atm and btw, lol"
cleaned_sample_text_acronyms = replace_acronyms(sample_text_with_acronyms)

print("Original Text:", sample_text_with_acronyms)
print("Cleaned Text:", cleaned_sample_text_acronyms)

Original Text: fyi, the meeting is atm and btw, lol
Cleaned Text: For Your Information, the meeting is At The Moment and By The Way, Laughing Out Loud


In [43]:
X_train_no_acronyms = X_train_no_emojis.apply(replace_acronyms)
X_test_no_acronyms = X_test_no_emojis.apply(replace_acronyms)

print("Original X_train_no_emojis head (first 5 entries):")
display(X_train_no_emojis.head())

print("\nProcessed X_train_no_acronyms head (first 5 entries):")
display(X_train_no_acronyms.head())

print("\nOriginal X_test_no_emojis head (first 5 entries):")
display(X_test_no_emojis.head())

print("\nProcessed X_test_no_acronyms head (first 5 entries):")
display(X_test_no_acronyms.head())

Original X_train_no_emojis head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operator need learn use device multiple level ...
289     company expects net sale half 2008 reach previ...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Processed X_train_no_acronyms head (first 5 entries):


432     russia raisin s food division s home market st...
3071    operator need learn use device multiple level ...
289     company expects net sale half 2008 reach previ...
2588    bridge 12 long located anasmotet road end new ...
4121    nokia captor announced resident evil degenerat...
Name: Text, dtype: object


Original X_test_no_emojis head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma medium s et...
Name: Text, dtype: object


Processed X_test_no_acronyms head (first 5 entries):


3207    company supposed deliver machinery veneer toms...
1684    una charlotte deploy ssh tactic connector enab...
1044    2009 lee man combined annual production capaci...
4145    s high figure european scale noon said recalli...
1538    finland corresponding service alma medium s et...
Name: Text, dtype: object

## Feature Engineering: Bag-of-Words (BoW)

We will now convert the preprocessed text data into numerical feature vectors using the Bag-of-Words model. This involves tokenizing the text and counting the occurrences of each word, creating a matrix where each row represents a document and each column represents a word from the vocabulary.

In [44]:
from sklearn.feature_extraction.text import CountVectorizer

# Initialize CountVectorizer for Bag-of-Words
# We use the preprocessed text after removing acronyms
bow_vectorizer = CountVectorizer()

# Fit the vectorizer on the training data and transform it
X_train_bow = bow_vectorizer.fit_transform(X_train_no_acronyms)

# Transform the test data using the *fitted* vectorizer
X_test_bow = bow_vectorizer.transform(X_test_no_acronyms)

print("Shape of X_train_bow:", X_train_bow.shape)
print("Shape of X_test_bow:", X_test_bow.shape)

# Display the first few feature names to understand the vocabulary
print("\nFirst 10 features (words) from BoW vectorizer:")
display(bow_vectorizer.get_feature_names_out()[:10])

Shape of X_train_bow: (3876, 8123)
Shape of X_test_bow: (970, 8123)

First 10 features (words) from BoW vectorizer:


array(['00', '000', '000063', '0008', '001', '002', '0025', '003', '0030',
       '004'], dtype=object)

## Model Construction and Evaluation: Random Forest with Bag-of-Words

Now, we will train a RandomForestClassifier using the Bag-of-Words (BoW) feature representation. After training, the model's performance will be evaluated on the test set using various classification metrics.

In [45]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

# Initialize the RandomForestClassifier
# Using a fixed random_state for reproducibility
rf_bow = RandomForestClassifier(random_state=42)

# Train the model using the Bag-of-Words transformed training data
print("Training Random Forest model with Bag-of-Words features...")
rf_bow.fit(X_train_bow, y_train)
print("Training complete.")

# Make predictions on the BoW transformed test data
y_pred_bow = rf_bow.predict(X_test_bow)

# Evaluate the model
print("\n--- Evaluation for Random Forest (Bag-of-Words) ---")
print("Accuracy:", accuracy_score(y_test, y_pred_bow))
print("\nClassification Report:\n", classification_report(y_test, y_pred_bow))

Training Random Forest model with Bag-of-Words features...
Training complete.

--- Evaluation for Random Forest (Bag-of-Words) ---
Accuracy: 0.7484536082474227

Classification Report:
               precision    recall  f1-score   support

    negative       0.81      0.42      0.55       110
     neutral       0.74      0.96      0.83       571
    positive       0.77      0.46      0.58       289

    accuracy                           0.75       970
   macro avg       0.77      0.61      0.65       970
weighted avg       0.76      0.75      0.73       970



## Feature Engineering: TF-IDF (Term Frequency-Inverse Document Frequency)

Next, we will convert the preprocessed text data into numerical feature vectors using the TF-IDF model. This approach assigns weights to words based on their frequency in a document and their inverse frequency across the entire corpus, giving more importance to rare but significant words.

In [46]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Initialize TfidfVectorizer
# We use the preprocessed text after removing acronyms
tfidf_vectorizer = TfidfVectorizer()

# Fit the vectorizer on the training data and transform it
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train_no_acronyms)

# Transform the test data using the *fitted* vectorizer
X_test_tfidf = tfidf_vectorizer.transform(X_test_no_acronyms)

print("Shape of X_train_tfidf:", X_train_tfidf.shape)
print("Shape of X_test_tfidf:", X_test_tfidf.shape)

# Display the first few feature names to understand the vocabulary
print("\nFirst 10 features (words) from TF-IDF vectorizer:")
display(tfidf_vectorizer.get_feature_names_out()[:10])

Shape of X_train_tfidf: (3876, 8123)
Shape of X_test_tfidf: (970, 8123)

First 10 features (words) from TF-IDF vectorizer:


array(['00', '000', '000063', '0008', '001', '002', '0025', '003', '0030',
       '004'], dtype=object)

## Model Construction and Evaluation: Random Forest with TF-IDF

Now, we will train a RandomForestClassifier using the TF-IDF feature representation. After training, the model's performance will be evaluated on the test set using various classification metrics to compare against the BoW model.

In [47]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

# Initialize the RandomForestClassifier
# Using a fixed random_state for reproducibility
rf_tfidf = RandomForestClassifier(random_state=42)

# Train the model using the TF-IDF transformed training data
print("Training Random Forest model with TF-IDF features...")
rf_tfidf.fit(X_train_tfidf, y_train)
print("Training complete.")

# Make predictions on the TF-IDF transformed test data
y_pred_tfidf = rf_tfidf.predict(X_test_tfidf)

# Evaluate the model
print("\n--- Evaluation for Random Forest (TF-IDF) ---")
print("Accuracy:", accuracy_score(y_test, y_pred_tfidf))
print("\nClassification Report:\n", classification_report(y_test, y_pred_tfidf))

Training Random Forest model with TF-IDF features...
Training complete.

--- Evaluation for Random Forest (TF-IDF) ---
Accuracy: 0.7463917525773196

Classification Report:
               precision    recall  f1-score   support

    negative       0.78      0.36      0.50       110
     neutral       0.74      0.96      0.83       571
    positive       0.78      0.48      0.59       289

    accuracy                           0.75       970
   macro avg       0.77      0.60      0.64       970
weighted avg       0.75      0.75      0.72       970



## Comparative Analysis: Bag-of-Words vs. TF-IDF

We have successfully built and evaluated two Random Forest classifiers: one using Bag-of-Words (BoW) features and another using TF-IDF features. Let's compare their performance metrics.

### 1. Random Forest with Bag-of-Words

**Accuracy:** 0.7484536082474227

**Classification Report:**
```
               precision    recall  f1-score   support

    negative       0.81      0.42      0.55       110
     neutral       0.74      0.96      0.83       571
    positive       0.77      0.46      0.58       289

    accuracy                           0.75       970
   macro avg       0.77      0.61      0.65       970
weighted avg       0.76      0.75      0.73       970
```

### 2. Random Forest with TF-IDF

**Accuracy:** 0.7463917525773196

**Classification Report:**
```
               precision    recall  f1-score   support

    negative       0.78      0.36      0.50       110
     neutral       0.74      0.96      0.83       571
    positive       0.78      0.48      0.59       289

    accuracy                           0.75       970
   macro avg       0.77      0.60      0.64       970
weighted avg       0.75      0.75      0.72       970
```

### Comparative Findings

1.  **Overall Accuracy**: Both models achieved very similar overall accuracy scores. The BoW model slightly edged out the TF-IDF model (0.748 vs. 0.746), but the difference is marginal and likely not statistically significant.

2.  **Precision and Recall (Class-wise)**:
    *   **Negative Class**: The BoW model showed slightly higher precision (0.81 vs. 0.78) and recall (0.42 vs. 0.36) for the 'negative' class, resulting in a better F1-score (0.55 vs. 0.50). This suggests BoW was slightly better at identifying negative sentiments without too many false positives and capturing more of the actual negative instances.
    *   **Neutral Class**: Both models performed exceptionally well on the 'neutral' class, achieving high recall (0.96) and F1-scores (0.83), indicating they are very good at identifying neutral statements. Precision was also very similar (0.74 for both).
    *   **Positive Class**: The TF-IDF model had a slightly higher recall for the 'positive' class (0.48 vs. 0.46) and a very slightly higher F1-score (0.59 vs. 0.58), while precision was almost identical (0.78 vs. 0.77). This implies TF-IDF might have captured a few more true positive instances compared to BoW.

3.  **F1-Score (Macro and Weighted Averages)**: The macro and weighted average F1-scores are also very close, indicating a comparable overall balance between precision and recall across all classes for both methods.

### Conclusion

In this specific sentiment analysis task using a Random Forest classifier, both Bag-of-Words and TF-IDF performed very similarly. The slight differences in class-wise metrics suggest that:

*   **BoW** might have a minor advantage in identifying negative sentiment with higher confidence.
*   **TF-IDF** might be slightly better at recalling positive sentiments.

The 'neutral' class is consistently well-predicted by both. The dataset might not contain enough distinguishing features for one vectorization technique to significantly outperform the other, or the Random Forest model is robust enough to handle the nuances of both representations effectively. For practical purposes, either approach would yield comparable results with the current setup.